components/scream/src/share/grid/remap/coarsening_remapper.cpp

```diff
@@ -0,0 +1,951 @@
+    const auto& f_src    = m_src_fields[i];
+    const auto& f_ov_tgt = m_ov_tgt_fields[i];
+
+    // Dispatch kernel with the largest possible pack size
+    const auto& src_ap = f_src.get_header().get_alloc_properties();
+    const auto& ov_tgt_ap = f_ov_tgt.get_header().get_alloc_properties();
+    if (can_pack && src_ap.is_compatible<RPack<16>>() &&
+                    ov_tgt_ap.is_compatible<RPack<16>>()) {
+      local_mat_vec<16>(f_src,f_ov_tgt);
+    } else if (can_pack && src_ap.is_compatible<RPack<8>>() &&
+                           ov_tgt_ap.is_compatible<RPack<8>>()) {
+      local_mat_vec<8>(f_src,f_ov_tgt);
+    } else if (can_pack && src_ap.is_compatible<RPack<4>>() &&
+                           ov_tgt_ap.is_compatible<RPack<4>>()) {
+      local_mat_vec<4>(f_src,f_ov_tgt);
+    } else {
+      local_mat_vec<1>(f_src,f_ov_tgt);
+    }
+  }
+
+  // Pack, then fire off the sends
+  pack ();
+  if (not m_send_req.empty()) {
+    int ierr = MPI_Startall(m_send_req.size(),m_send_req_ptr);
+    EKAT_REQUIRE_MSG (ierr==MPI_SUCCESS,
+        "Error! Something whent wrong while starting persistent send requests.\n"
+        "  - send rank: " + std::to_string(m_comm.rank()) + "\n");
+  }
+
+  // Wait for all data to be received, then unpack
+  if (not m_recv_req.empty()) {
+    int ierr = MPI_Waitall(m_recv_req.size(),m_recv_req_ptr, MPI_STATUSES_IGNORE);
+    EKAT_REQUIRE_MSG (ierr==MPI_SUCCESS,
+        "Error! Something whent wrong while waiting on persistent recv requests.\n"
+        "  - recv rank: " + std::to_string(m_comm.rank()) + "\n");
+  }
+  unpack ();
+
+  // Wait for all sends to be completed
+  if (not m_send_req.empty()) {
+    int ierr = MPI_Waitall(m_send_req.size(),m_send_req_ptr, MPI_STATUSES_IGNORE);
+    EKAT_REQUIRE_MSG (ierr==MPI_SUCCESS,
+        "Error! Something whent wrong while waiting on persistent send requests.\n"
+        "  - send rank: " + std::to_string(m_comm.rank()) + "\n");
+  }
+
+}
+
+template<int PackSize>
+void CoarseningRemapper::
+local_mat_vec (const Field& x, const Field& y) const
+{
+  using RangePolicy = typename KT::RangePolicy;
+  using MemberType  = typename KT::MemberType;
+  using ESU         = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+  using Pack        = ekat::Pack<Real,PackSize>;
+  using PackInfo    = ekat::PackInfo<PackSize>;
+
+  const auto& src_layout = x.get_header().get_identifier().get_layout();
+  const int rank = src_layout.rank();
+  const int nnz = m_weights.size();
+  auto row_col_lids = m_row_col_lids;
+  auto weights = m_weights;
+  switch (rank) {
+    case 1:
+    {
+      auto x_view = x.get_view<const Real*>();
+      auto y_view = y.get_view<      Real*>();
+      Kokkos::parallel_for(RangePolicy(0,nnz),
+                           KOKKOS_LAMBDA(const int& i) {
+        const auto row = row_col_lids(i,0);
+        const auto col = row_col_lids(i,1);
+        const auto w = weights(i);
+        y_view(row) += w*x_view(col);
```

Comment:
On GPU this is a write race condition on `y_view(row)`. For this reason, during init, I recommend formatting the local matrices as CRS. Then you can do the ||4 over rows rather than nonzeros.

Reply:
Ah, good point.